### PR TITLE
ES Recommendation Engine added categories to items to look for and increased ratings value to 3.5

### PR DESCRIPTION
**Pull request by seenusuvarna**
Please Review the changes for the changes that have added categories to the items being recommended.

The Categories are determined by the recommended and bookmarked listings in a user profile and then Elasticsearch uses those items to find users with similar listings in their profiles.  It then ranks the listings based on reviews of greater than 3.5 to return the results.

To test:  (SEE UPDATE AT END)
Run make recommend_es_user
Use bettafish user and a list of listings should come back for the recommended list then review a couple of the apps that are not in the recommended list and give high ratings > 3.  Run make recommend_es_user again
Look at the results and should notice apps from the same category appear.
(Ex. No Jobspot was showing up, so I reviewed Jobspot 1 and Jpbspot 9 and gave a review of 5.  I then reran the make recommend_es_user and it now shows Jobspot apps in my recommended list)

Please review changes that were done to the Elasticsearch User Based Recommendation System:
@mannyrivera2010 
@skhtet 
@blynch11p 
@J-Fid 
@mleeBoeing 

**UPDATED TESTING INFORMATION:**
The following should help document the behavior of how the Elasticsearch recommendation engine should work with the changes that have been done.
First start off with a clean environment by running: make dev
Then run: make recommend_es_user
Start the back-end and front-end with Elasticsearch running and see the recommended apps as follows:
(The first three should be the same, but the remaining ones may be different due to jittering)
Possible Initial list (May be different due to jittering):
- Bread Basket
- Skybox 1
- Location Lister 4
- Location Lister 5
- Location Lister 6
- Location Lister 8
- Air Mail
- Skybox 2
- Skybox 3
- JotSpot 1

Another sample:
- Bread Basket
- Skybox 1
- LocationLister 4
- LocationLister 5
- LocaitonLister 6
- LocationLister 8
- LocationLister 9
- Chart Course
- Air Mail

Then Bookmark the following Apps:
JotSpot 5
JotSpot 6
(Note what categories they are in as well)
LEAVE Web Window Open with current recommendation listings

Stop Backend and run: make recommend_es_user
Start Backend again

On a different window open up Home page site.
Results should look like:
- JotSpot 1
- JotSpot 2
- JotSpot 3
- JotSpot 4
- JotSpot 8
- JotSpot 9-
- LocationLister 1
- LocationLister 2
- LocationLister 5
- LocationLister 7

Another sample:
- JotSpot 1
- JotSpot 2
- JotSpot 3
- JotSpot 8
- JotSpot 9
- LocationLister
- LocationLister 1
- LocationLister 2
- LocationLister 3
- LocationLister 4
(Note the categories that the listings are in and should be related to JotSpot 5 and 6)

The listings should now have items related to the categories that were bookmarked.  Note that if items are reviewed with a ranking of 1-3, then the categories are not included in the ones that are analyzed.

### CODE REVIEW

File: ozpcenter/recommend/recommend.py [hunk 1]
```diff
@@ -498,61 +571,106 @@ def recommendation_logic(self):
...
-                                    {"terms": {"bookmark_ids": bookmarked_list}},
-                                    {
-                                        "nested": {
-                                            "path": "ratings",
-                                            "query": {
-                                                "bool": {
-                                                    "should": [
-                                                        {"terms": {"ratings.listing_id": bookmarked_list}}
-                                                    ]
+                if len(categories) > 0:
```
Comment (blynch11p) [inline ozpcenter/recommend/recommend.py]: I see that `if len(categories) > 0:` is used in both the parent if and else blocks for bookmarks with the difference being the inclusion or exclusion of `{"terms": {"bookmark_ids": bookmarked_list}}`
Is there a way to refactor in a separate if/else dedicated for the categories and build upon the json object `agg_query_term` after the bookmarks blocks are complete?

Not a requirement to change, just a suggestion/question.

Comment (seenusuvarna) [inline ozpcenter/recommend/recommend.py]: This is a good suggestion.  I did it for readability, but think that your idea is one that I should implement.  I will look into implementing your suggestion.

Comment (seenusuvarna) [inline ozpcenter/recommend/recommend.py]: Made changes and need to do some more checks on it to make sure that it works as expected.  Will submit for review after testing completed.

**Comment by seenusuvarna** [comment]
Please review changes after comments were made in addition to original changes.
@blynch11p 
@mannyrivera2010 
@skhtet 
@mleeBoeing 
@J-Fid

**Comment by blynch11p** [comment]
Using your example above (See no JotSpot, review some JotSpot with 5 star reviews, rerun `make recommend_es_user`) I still do not see JotSpots.  Does it matter what user is doing the reviewing?

**Comment by blynch11p** [comment]
Pt2: I used another user and also gave those JotSpots 5 star reviews. No JotSpot. I then bookmarked everything in my Recommended list.  JotSpot never showed up and now I have no Recommendations at all.

**Comment by seenusuvarna** [comment]
Brian, after changing the code and correcting a missing item that I caught, the behavior has changed.  I will get the correct behavior and let everyone know what should happen.  I will say that if you un-comment the code ("Key {}, Score {}") under:
for indexitem in recommended_items:

Then if you run the recommendation before and after reviewing you will see the weights change. The exact behavior will be documented when I come in tomorrow.  I am also adding documentation to the code base.

**Comment by mannyrivera2010** [comment]
:-1:   It seems to be an issue with merging master into this branch

**Comment by seenusuvarna** [comment]
@mannyrivera2010 must have happened when I was trying to squash the commits.  I will redo the changes and submit a new review.  Thanks for catching it.  I checked earlier and did not see it.

**Comment by seenusuvarna** [comment]
Updated code with comments, updated query information and added testing information to help in understanding the behavior.
@blynch11p
@mannyrivera2010
@skhtet
@mleeBoeing
@J-Fid

**Comment by mannyrivera2010** [comment]
:+1:

**Comment by skhtet** [comment]
Seems fine for now (yet to be tested on staging onwards) :+1:

**Comment by blynch11p** [comment]
Pulled code.  Tested using the steps above and did correctly see more JotSpot recommendations in my list.
Additionally I reviewed some items and had other items from similar categories appear.
I reviewed the same items using different users to add a high number of reviews, and those apps appeared higher in the recommendation list.

:+1:

**Pull request merged**